core/Pos/Crypto/HD.hs

```diff
@@ -0,0 +1,103 @@
+import           Universum
+
+import           Pos.Binary.Class             (Bi (..))
+import           Pos.Crypto.Signing           (PublicKey (..), SecretKey (..))
+
+-- | Passphrase is a hash of root public key.
+-- We don't use root public key for store money, we use hash of it instead.
+newtype HDPassphrase = HDPassphrase ByteString
+
+-- | HDAddressPayload consists of
+-- * serialiazed and encrypted with symmetric scheme with passphrase (via ChaChaPoly1305 algorithm)
+-- * cryptographic tag
+-- For more information see 'packHDAddressAttr' and 'encryptChaChaPoly'.
+newtype HDAddressPayload = HDAddressPayload ByteString
+    deriving (Eq, Ord, Show, NFData, Generic)
+
+-- | Compute passphrase as hash of the root public key.
+deriveHDPassphrase :: PublicKey -> HDPassphrase
+deriveHDPassphrase (PublicKey pk) = HDPassphrase $
+    PBKDF2.generate
+        (PBKDF2.prfHMAC SHA512)
+        (PBKDF2.Parameters 500 passLen)
+        (unXPub pk)
+        ("address-hashing"::ByteString)
+  where
+    -- Password length in bytes
+    passLen = 32
+
+-- Direct children of node are numbered from 0 to 2^32-1.
+-- Child with index less or equal @maxHardened@ is a hardened child.
+maxHardened :: Word32
+maxHardened = fromIntegral $ (2::Word32)^(31::Word32)-1
+
+-- | Derive public key from public key via non-hardened (normal) way.
+-- If you try to pass index more than @maxHardened@, error will be called.
+deriveHDPublicKey :: PublicKey -> Word32 -> PublicKey
+deriveHDPublicKey (PublicKey xpub) childIndex
+    -- Is it the best solution?
+    | childIndex <= maxHardened = error "Wrong index for non-hardened derivation"
+    | otherwise = PublicKey $ deriveXPub xpub (childIndex - maxHardened - 1)
+
+-- | Derive secret key from secret key.
+-- If @childIndex@ <= @maxHardened@ key will be deriving hardened way, otherwise non-hardened.
+deriveHDSecretKey :: ByteArrayAccess passPhrase
+                  => passPhrase
+                  -> SecretKey
+                  -> Word32
+                  -> SecretKey
+deriveHDSecretKey passPhrase (SecretKey xprv) childIndex
+  | childIndex <= maxHardened =
+      SecretKey $ deriveXPrvHardened passPhrase xprv childIndex
+  | otherwise =
+      SecretKey $ deriveXPrv passPhrase xprv (childIndex - maxHardened - 1)
+
+-- | Serialize tree path and encrypt it using passphrase via ChaChaPoly1305.
+packHDAddressAttr :: HDPassphrase -> [Word32] -> HDAddressPayload
+packHDAddressAttr (HDPassphrase simkey) path = do
+    let pathSer = BSL.toStrict $ runPut (put path)
+    let packCF =
+          encryptChaChaPoly
+              "serokellfore"
+              simkey
+              "addressattr" -- I don't know what does mean this parameter.
+                            -- Seems value isn't important...
+              pathSer
+    case packCF of
+        CryptoFailed er -> error $ "Error during packHDAddressAttr: " <> show er
+        CryptoPassed p  -> HDAddressPayload p
+
+-- Wrapper around ChaChaPoly1305 module.
+encryptChaChaPoly
```

Comment:
How can i be sure it's adequate? Can't find anything similar in sources (https://hackage.haskell.org/package/cryptonite-0.22/docs/src/Crypto-Cipher-ChaChaPoly1305.html#encrypt).